uavcan/register/384.Access.1.0.uavcan

```diff
@@ -15,33 +18,66 @@
 # case properly.
 #
 # The timestamp provided in the response corresponds to the time when the register was read. The timestamp may
-# be empty if the server does not support timestamping or its clock is not yet synchronized with the bus.
+# be empty if the server does not support timestamping or its clock is not (yet) synchronized with the network.
 #
-# If only read is desired, but not write, the caller shall provide a value of type Empty. That will signal the server
+# If only read is desired, but not write, the caller shall provide a value of type 'empty'. That will signal the server
 # that the write operation shall be skipped, and it will proceed to read the register immediately.
 #
 # If the requested register does not exist, the write operation will have no effect and the returned value will be
-# empty. Existing registers should not return Empty when read since that would make them indistinguishable from
+# empty. Existing registers should not return 'empty' when read since that would make them indistinguishable from
 # nonexistent registers.
 #
+#
+#   REGISTER DEFINITION REQUIREMENTS
+#
 # Registers shall never change their type or flags as long as the server is running. Meaning that:
 #   - Mutability and persistence flags cannot change their states.
 #   - Read operations shall always return values of the same type and same dimensionality.
 #     The dimensionality requirement does not apply to inherently variable-length values such as strings and
 #     unstructured chunks.
 #
-# In order to discover the type of a register, the caller needs to invoke this service with the write request set
-# to Empty. The response will contain the current value of the register with the type information (which doesn't
-# change).
+# Register name should contain only:
+#   - Lowercase ASCII alphanumeric characters (a-z, 0-9)
+#   - Full stop (.)
+#   - Low line (underscore) (_)
+# With the following limitations/recommendations:
+#   - The name shall not begin with a decimal digit (0-9).
+#   - The name shall neither begin nor end with a full stop.
+#   - A low line shall not be followed by a non-alphanumeric character.
+#   - The name should contain at least one full stop character.
+# Other patterns and ASCII characters are reserved for special function registers (introduced below).
+#
+#
+#   ENVIRONMENT VARIABLES
```

Comment:
I really hate encouraging environment variables as a primary mechanism. It's fine as a secondary but I'd rather we define a configuration file standard as the primary mechanism. Can we not just use the same mapping but encourage a file of `KEY=VALUE\n` as the way to specify each?

Reply:
Environment variables are not meant to be directly used by the human integrator/developer excepting some unusual circumstances that require low-level access. They are convenient because they are present virtually everywhere and are familiar to virtually everybody, so it is a solid foundation to build upon. What you are describing can be implemented *on top* of the environment variable conventions that we define here. For instance, the configuration file-like behavior can be obtained using shell scripts or the orchestrator that you are already familiar with.

```sh
# my-configuration-file
UAVCAN__NODE__ID=42
UAVCAN__UDP__IFACE=127.9.0.0
UAVCAN__SUB__TEMPERATURE_SETPOINT__ID=2345
UAVCAN__SUB__TEMPERATURE_MEASUREMENT__ID=2346
UAVCAN__PUB__HEATER_VOLTAGE__ID=2347
UAVCAN__SRV__LEAST_SQUARES__ID=123
UAVCAN__DIAGNOSTIC__SEVERITY=2
```

```
$ . my-configuration-file
$ ./run_this
```

Reply:
environment variables infer a shell environment. We are creating a dependency on the presence of a shell.

Reply:
Yes, this is by design:

> This section applies only to software nodes executed in a high-level operating system that supports environment variables or an equivalent mechanism.

A deeply embedded system is unlikely to source its configuration from a shell or a generic configuration file, so it doesn't seem relevant for that use case.

Reply:
I don't know what to say. I think it's a bad design choice. You can override my opinion if you disagree.

Reply:
But if shell dependency is intentional and user-friendliness is not relevant, what makes it bad?

Reply:
1. It's an abuse of environment variables since you are basically treating it as a list. If anything it should be a delineated value within a single environment variable to be consistent with things like $PATH
2. Not all build automation tools support environment variables but they all support files.

Reply:
1. The objective is to assign registers selectively such that each register is treated as an independent entity. A node process that is unaware of register X would not read the corresponding environment variable, so unless I misunderstood what you mean we don't seem to be following a list-like semantics here.

2. This is meant to be a method of integration-time/runtime configuration management. What build tools are you referring to and why is this relevant?